clients/drcachesim/tracer/raw2trace.h

```diff
@@ -1073,33 +1080,66 @@ template <typename T> class trace_converter_t {
                 (instr->reads_memory() || instr->writes_memory()) &&
                 // No following memref for instruction-only trace type.
                 !is_instr_only_trace) {
-                for (uint j = 0; j < instr->num_mem_srcs(); j++) {
-                    error = append_memref(tls, &buf, instr, instr->mem_src_at(j), false,
-                                          reg_vals);
-                    if (!error.empty())
-                        return error;
-                    error = handle_kernel_interrupt_and_markers(
-                        tls, &buf, cur_modoffs, instr->length(), instrs_are_separate,
-                        &interrupted);
-                    if (!error.empty())
-                        return error;
-                    if (interrupted)
-                        break;
-                }
-                // We break before subsequent memrefs on an interrupt, though with
-                // today's tracer that will never happen (i#3958).
-                for (uint j = 0; !interrupted && j < instr->num_mem_dests(); j++) {
-                    error = append_memref(tls, &buf, instr, instr->mem_dest_at(j), true,
-                                          reg_vals);
-                    if (!error.empty())
-                        return error;
-                    error = handle_kernel_interrupt_and_markers(
-                        tls, &buf, cur_modoffs, instr->length(), instrs_are_separate,
-                        &interrupted);
-                    if (!error.empty())
-                        return error;
-                    if (interrupted)
-                        break;
+                if (instr->is_scatter_or_gather()) {
+                    // The instr should either load or store, but not both. Also,
+                    // it should have a single src or dest operand.
+                    DR_ASSERT(instr->num_mem_srcs() + instr->num_mem_dests() == 1);
+                    bool is_scatter = instr->num_mem_dests() == 1;
+                    bool reached_end_of_memrefs = false;
+                    // For expanded scatter/gather instrs, we do not have prior knowledge
+                    // of the number of store/load memrefs that will be present. So we
+                    // continue reading entries until we find a non-memref entry.
```

Comment:
Comment that this only works because the scatter/gather is the only instruction in the block: we assume that drx_expand_* isolates it to a block by itself.

Reply:
But this logic (of scanning memrefs following a scatter/gather instr until we reach a non-memref) doesn't really depend on scatter/gather being in its own block, does it? The important thing is that we can never have memrefs on their own, without being preceded by a pc entry, which is true always. So even if there were other app instrs following/before the scatter/gather expansion in the same block, the logic would still work.

Reply:
> The important thing is that we can never have memrefs on their own, without being preceded by a pc entry, which is true always.

No, this is not true.  I think you are conflating raw traces with final traces.  Raw traces have a single PC entry for the whole block, followed by memref entries for all the instructions in the blcok, without any more PC entries in between.  That's why raw2trace has to go figure out what memref goes with which instruction.

Reply:
I think I was. I understand now. Added  the comment about separate bb being important for this logic to work.